File: src/0215-Kth-Largest-Element-in-an-Array/0215.py
class Solution:
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        nums_len = len(nums)
        l = 0
        r = nums_len - 1
        while l < r:
            tl = l
            tr = r
            tmp = nums[tl]
            while tl < tr:
                while tr > tl and nums[tr] > tmp:
                    tr -= 1
                if tr > tl:
                    nums[tl] = nums[tr]
                    tl += 1
                while tl < tr and nums[tl] <= tmp:
                    tl += 1
                if tl < tr:
                    nums[tr] = nums[tl]
                    tr -= 1
                
            nums[tl] = tmp
            if tl == nums_len - k:
                return tmp
            elif tl < nums_len - k:
                l = tl + 1
            else:
                r = tl - 1

        return nums[l]
        
if __name__ == "__main__":
    nums = [3,2,1,5,6,4]
    k = 2
    print(Solution().findKthLargest(nums, k))

File: src/0215-Kth-Largest-Element-in-an-Array/0215.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <functional>
using namespace std;

static int x = []() {std::ios::sync_with_stdio(false); cin.tie(0); return 0; }();
class Solution 
{
public:
    int findKthLargest(vector<int>& nums, int k) 
    {
        int l = 0, r = nums.size() - 1;
        while (true)
        {
            int p = _partition(nums, l, r);
            if (p == k - 1) return nums[p];
            if (p > k - 1) r = p - 1;
            else l = p + 1;
        }
    }
private:
    int _partition(vector<int>& nums, int l, int r)
    {
        int j = (rand() % (r - l + 1)) + l;
        swap(nums[l], nums[j]);
        j = l;
        for (int i = l + 1; i <= r; ++i)
        {
            if (nums[i] > nums[l])
            {
                ++j;
                swap(nums[i], nums[j]);
            }
        }
        swap(nums[l], nums[j]);
        return j;
    }
};
int main()
{    
    vector<int> nums = {3,2,3,1,2,4,5,5,6};
    int k = 4;
    cout << Solution().findKthLargest(nums, k);
}